tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,529 @@
+            (std::ptr::null_mut() as *const u64, 0u64)
+        } else {
+            let offsets = offsets.unwrap();
+            (offsets.as_ptr(), std::mem::size_of_val(offsets) as u64)
+        };
+
+        // An important note here is that the Enumeration allocator copies the
+        // contents of data and offsets rather than assumes ownership. That
+        // means this is safe as those bytes are guaranteed to be alive until
+        // we drop self at the end of this method after returning from
+        // tiledb_enumeration_alloc.
+        let res = unsafe {
+            ffi::tiledb_enumeration_alloc(
+                self.context.capi(),
+                c_name.as_c_str().as_ptr(),
+                c_dtype,
+                self.cell_val_num,
+                if self.ordered { 1 } else { 0 },
+                data.as_ptr() as *const std::ffi::c_void,
+                std::mem::size_of_val(data) as u64,
+                offsets_ptr as *const std::ffi::c_void,
+                offsets_len,
+                &mut c_enmr,
+            )
+        };
+
+        if res == ffi::TILEDB_OK {
+            Ok(Enumeration {
+                context: self.context,
+                raw: RawEnumeration::Owned(c_enmr),
+            })
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+}
+
+/// Encapsulation of data needed to construct an Enumeration
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct EnumerationData {
+    pub name: String,
+    pub datatype: Datatype,
+    pub cell_val_num: Option<u32>,
+    pub ordered: Option<bool>,
+    pub data: Box<[u8]>,
+    pub offsets: Option<Box<[u64]>>,
+}
+
+impl<'ctx> TryFrom<&Enumeration<'ctx>> for EnumerationData {
+    type Error = crate::error::Error;
+
+    fn try_from(enmr: &Enumeration<'ctx>) -> TileDBResult<Self> {
+        let datatype = enmr.datatype()?;
+        let cell_val_num = enmr.cell_val_num()?;
+        let data = Box::from(enmr.data()?);
+        let offsets: Option<Box<[u64]>> = enmr.offsets()?.map(Box::from);
+
+        Ok(EnumerationData {
+            name: enmr.name()?,
+            datatype,
+            cell_val_num: Some(cell_val_num),
+            ordered: Some(enmr.ordered()?),
+            data,
+            offsets,
+        })
+    }
+}
+
+impl<'ctx> Factory<'ctx> for EnumerationData {
+    type Item = Enumeration<'ctx>;
+
+    fn create(&self, context: &'ctx Context) -> TileDBResult<Self::Item> {
+        let mut b = Builder::new(context, &self.name, self.datatype);
+
+        if let Some(cvn) = self.cell_val_num {
+            b = b.cell_val_num(cvn);
+        }
+
+        if let Some(ordered) = self.ordered {
+            b = b.ordered(ordered);
+        }
+
+        let offsets: Option<&[u64]> = if let Some(offsets) = &self.offsets {
+            Some(&offsets[..])
+        } else {
+            None
+        };
+
+        b.build(&self.data[..], offsets)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'd suggest also a `test_eq` which confirms that changing any of the fields makes the equality false.

Reply:
Done.